Minimal Hamiltonian Path (#4)
golemfactory/wasm-store: MinimalHamiltonianPath/README.md

The program searches for minimal Hamiltonian path

> in direct weighted graphs.

Reviewer: ```suggestion
in weighted directed graphs.
```

### Input file

The file `in.txt` contains graph. It is in the form of a table of integers with `n` rows and `n` columns. A field `(i,j)` gives the weight of the direct edge from `i`-th node to `j`-th node. Negative value means that the edge does not exists. All diagonal fields must be negative. Separator must be a single space.

### Execution and output

You should run multiple instances of the program. Each takes the same input file but different parameters. Each instance writes answer to the output file `out.txt`, so watch out to not override it. The output file contains the minimal Hamiltonian path the the instance found. In the first row is total weight of the path, in the second row are nodes in the order that forms the path. Negative total weight means that the instance did not find the path.

### Parameters

> The program takes two parameters. The first is a reminder `r`, the second is denominator `d`. The program searches all paths starting from any `i`-th node such that `i mod d = r`. So if you want to search Hamiltonian paths with parallel degree of `3`, you should run `3` instances with parameters `0 3`, `1 3`, `2 3`.

Reviewer: ```suggestion
The program takes two parameters. The first is a remainder `r`, the second is denominator `d`. The program searches all paths starting from any `i`-th node such that `i mod d = r`. So if you want to search Hamiltonian paths with parallel degree of `3`, you should run `3` instances with parameters `0 3`, `1 3`, `2 3`.
```

### Algoritm

> The algorithm is brute force. It searches all possible paths in DFS manner.

Reviewer: ```suggestion
The algorithm is naive. It searches all possible paths in DFS manner.
```

### Running

At first copy the directory `MinimalHamiltonianPath` to your local drive and go to this directory. And make sure that your Golem node is up and running.

Before running the example, edit the file `task.json` and update fields `input_dir` and `output_dir`.

> If you makes any changes to `mhp.c` file, do the following.

Reviewer: ```suggestion
If you make any changes to `mhp.c` file, do the following.
```

```
emcc -o mhp.js -s BINARYEN_ASYNC_COMPILATION=0 mhp.c
mv mhp.js in/
mv mhp.wasm in/
```

> If you want to run the algorithm with the parallel degree of `3` for instance, you need to have the proper subdirectories in `in` directory and configured `task.json`. 

Reviewer: I'm guessing that "parallel degree of `3`" refers to creating 3 Golem subtasks, correct?

Author: correct

Reviewer: Can we state it like that explicitly then? So instead of saying "parallel degree of 3" simply say "execute 3 Golem tasks in parallel"? I find the former a bit confusing.

Author: i added it in braces
